oeps/architectural-decisions/oep-0065-frontend-composability.rst

```diff
@@ -0,0 +1,370 @@
+
+Runtime module loading improves:
+
+- Frequency of full page refreshes
+- Composability across MFEs/domains
+- Runtime extensiblity
+
+It also helps preserve the independent deployability and autonomy of our MFEs.
+
+Shared Dependencies
+===================
+
+Going along with runtime module loading is the ability to share dependencies between MFEs. This is another core capability of Webpack module federation.
+
+As noted above, MFEs have a number of dependencies which are common between them, but which aren't actually shared at runtime in any way. The capability to share these dependencies - such as ``react``, ``@openedx/paragon``, etc. - would mitigate a great deal of our inconsistency and UX/DX issues.
+
+We expect the following packages - which are used in the vast majority of MFEs today - could all be shared between MFEs.
+
+.. list-table::
+   :widths: 30 20 50
+
+   * - **Package**
+     - **Estimated Size**
+     - **Notes**
+   * - @edx/brand
+     - Variable
+     -
+   * - @edx/frontend-component-footer
+     - 88.1k
+     -
+   * - @edx/frontend-component-header
+     - 156.9k
+     -
+   * - @edx/frontend-platform
+     - 355.3k
+     -
+   * - @openedx/paragon (or @edx/paragon)
+     - ~950k
+     - bundlephobia.com threw an error on the Paragon build, but we expect it’s the largest package here.
+   * - classnames
+     - 0.8k
+     -
+   * - core-js
+     - 241.1k
+     -
+   * - prop-types
+     - 0.9k
+     -
+   * - react
+     - 6.4k
+     -
+   * - react-dom
+     - 130.2k
+     -
+   * - react-redux
+     - 11.2k
+     -
+   * - react-router
+     - 58.9k
+     -
+   * - react-router-dom
+     - 77.1k
+     -
+   * - redux
+     - 3.7k
+     - Ultimately should be replaced with @reduxjs/toolkit. This document isn't making a judgment on the utility of redux, just noting that it is currently in use in the vast majority of MFEs.
+   * - regenerator-runtime
+     - 6.6k
+     -
+
+Total size: ~2,087.2k unzipped
+
+Sharing these dependencies has several important positive effects:
+
+- Enforces runtime consistency across all types of dependencies
+- Reduces runtime bundle size
+
+Note that "build time" and "dependency maintenance" are not mentioned above.
+
+Interestingly, because of the way Webpack module federation's dependency sharing is implemented, this may not have a large impact on build time or dependency maintenance. "Guest" MFEs still need to have the dependencies installed for the build to succeed, and artifacts for them will still be created and deployed - they just won't generally be used at runtime unless there's a version conflict.
+
+These dependencies still need to exist in the package.json file (and be installed by npm) for the build to succeed, and for IDE hinting, auto completion, linting, etc.
```

Comment:
> artifacts for them will still be created and deployed - they just won't generally be used at runtime unless there's a version conflict.

It's not obvious to me how this works. I assume there are docs that explain this somewhere, maybe linking to those here would be helpful?

Reply:
Yep, and this seems to conflict with what's below.

To me, "they just won't generally be used at runtime unless there's a version conflict." implies that every MFE _can_ use different versions of dependencies and they'll be de-duplicated in a smart way - if MFE A uses Paragon 25, it gets Paragon 25, while if MFEs B and C both use Paragon 24.x.x (say, different but compatible minor releases), then MFEs B and C will transparently use a common Paragon module at runtime. I think that sounds great.

But the rest of this text seems to imply "we have to force all of our MFEs to keep their dependencies versioned in lock-step or module federation won't work" which sounds less than ideal.

Reply:
It is touched on at the very end

https://github.com/openedx/open-edx-proposals/blob/7f887a97ed345642ba8ed198258ff3a662718186/oeps/architectural-decisions/oep-0065-frontend-composability.rst?plain=1#L350-L351

but I think touching on it earlier and linking out to docs with more details would help provide context

Reply:
I've tried to simplify and clarify this section, and reference further down in the doc where there's a link to more details.  It may be that it still needs a diagram to be clear; let me know what y'all think.